core/src/main/clojure/xtdb/types.clj

```diff
@@ -1230,21 +1337,48 @@
     (cond
       (= 1 (count col-types)) (first col-types)
       (set/subset? col-types #{:float4 :float8}) :float8
-      (set/subset? col-types #{:int2 :int4 :int8}) :int8)))
+      (set/subset? col-types #{:int2 :int4 :int8}) :int8
+      (is-list-type col-types) [:list (->unified-col-type (second col-types))])))
+
+(defn remove-nulls
+  [typ]
+  (if (is-list-type typ)
+    (let [inner (-> typ
+                    second
+                    (disj :null))]
+      ;; wrap in extra vector as next step maps - we want the whole [:list :inner] evaluated
+      [[:list (if (= 1 (count inner))
+                (first inner)
+                inner)]])
+    (disj typ :null)))
 
 (defn field->pg-type
   ([field] (field->pg-type nil field))
 
   ([fallback-pg-type ^Field field]
    (let [field-name (.getName field)
          col-type (field->col-type field)
-         col-types (-> (flatten-union-types col-type)
-                       (disj :null)
+         col-types (-> ((if (is-list-type col-type)
+                          flatten-list-unions
+                          flatten-union-types) col-type)
+                       remove-nulls
```

Comment:
I find this together with the above functions hard to parse and understand. Especially the ignoring of lists seems incorrect. I understand that you want to get rid of inner nulls. Could we maybe have:
- a function `remove-nulls` that takes an arbitrary col-type and removes any nulls (even arbitrary nested ones)
- together with `flatten-union-types` you remove unnessary unions and then only in the `->unified-col-type` code path you match against an "pure" pgwire arrya

Reply:
We can also pair if that would help.

Reply:
yeah, definitely - let's review this - if it's hard to parse - definitely needs improving 👍

Reply:
Resolved offline.